metricflow/dataflow/optimizer/source_scan/source_scan_optimizer.py

```diff
@@ -0,0 +1,368 @@
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_base_output_handler(node)
+
+    def visit_join_aggregated_measures_by_groupby_columns_node(  # noqa: D
+        self, node: JoinAggregatedMeasuresByGroupByColumnsNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_base_output_handler(node)
+
+    def visit_aggregate_measures_node(  # noqa: D
+        self, node: AggregateMeasuresNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_base_output_handler(node)
+
+    def visit_compute_metrics_node(  # noqa: D
+        self, node: ComputeMetricsNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        # Run the optimizer on the parent branch to handle derived metrics, which are defined recursively in the DAG.
+        optimized_parent_result: OptimizeBranchResult = node.parent_node.accept(self)
+        if optimized_parent_result.base_output_node is not None:
+            return OptimizeBranchResult[SourceDataSetT](
+                base_output_node=ComputeMetricsNode(
+                    parent_node=optimized_parent_result.base_output_node,
+                    metric_specs=node.metric_specs,
+                )
+            )
+
+        return OptimizeBranchResult(base_output_node=node)
+
+    def visit_order_by_limit_node(  # noqa: D
+        self, node: OrderByLimitNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_base_output_handler(node)
+
+    def visit_where_constraint_node(  # noqa: D
+        self, node: WhereConstraintNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_base_output_handler(node)
+
+    def visit_write_to_result_dataframe_node(  # noqa: D
+        self, node: WriteToResultDataframeNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_sink_node_handler(node)
+
+    def visit_write_to_result_table_node(  # noqa: D
+        self, node: WriteToResultTableNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_sink_node_handler(node)
+
+    def visit_pass_elements_filter_node(  # noqa: D
+        self, node: FilterElementsNode[SourceDataSetT]
+    ) -> OptimizeBranchResult[SourceDataSetT]:
+        self._log_visit_node_type(node)
+        return self._default_base_output_handler(node)
+
+    @staticmethod
+    def _combine_branches(
+        left_branches: Sequence[BaseOutput[SourceDataSetT]], right_branch: BaseOutput[SourceDataSetT]
+    ) -> Sequence[BranchCombinationResult]:
```

Comment:
from our chat earlier, comment here to explain the assumption of how this is optimal would be nice!